### PR TITLE
Add a test for the ABF element in 3d.

### LINKED ISSUE #2785: Initializing FE_ABF fails
While trying to initialize FE_ABF through FETools::get_fe_by_name (PR #2784),
I noticed that we fail for all polynomial degrees with errors like

---

An error occurred in line <99> of file <../source/fe/fe.cc> in function
     dealii::FiniteElement<<anonymous>, <anonymous> >::FiniteElement(const dealii::FiniteElementData<dim>&, const std::vector<bool>&, const std::vectordealii::ComponentMask&) [with int dim = 3; int spacedim = 3]
 The violated condition was: 
     restriction_is_additive_flags.size() == this->dofs_per_cell
 The name and call sequence of the exception was:
     ExcDimensionMismatch(restriction_is_additive_flags.size(), this->dofs_per_cell)
 Additional Information: 
 Dimension 36 not equal to 48

---

for FE_ABF<3>(1).

### PR DESCRIPTION
**Pull request by bangerth**
This closes #2785.

### CODE REVIEW

File: tests/fe/abf_02.cc [hunk 1]
```diff
@@ -0,0 +1,593 @@
...
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+
+
+// Check that we can generate an ABF(0) element in 3d
```
Comment (tjhei) [inline tests/fe/abf_02.cc]: can you expand this please? I think you are projecting a function and then checking the result, right?

File: tests/fe/abf_02.cc [hunk 2]
```diff
@@ -0,0 +1,593 @@
...
+
+      // Get values from solution vector (For Trap.Rule)
+      std::vector<std::vector<Tensor<1,3> > >
+      grads_here (n_q_points, std::vector<Tensor<1,3> > (n_components));
+      fe_values.get_function_gradients (solution, grads_here);
+
+      for (unsigned int q_point=0; q_point<n_q_points; ++q_point)
+        {
+          //    double u0 = this_value[q_point](0);
+          //double v0 = this_value[q_point](1);
```
Comment (tjhei) [inline tests/fe/abf_02.cc]: remove

File: tests/fe/abf_02.cc [hunk 3]
```diff
@@ -0,0 +1,593 @@
...
+          double u0 = 0;
+          double v0 = 0;
+          for (unsigned int i = 0; i < dofs_per_cell; ++i)
+            {
+              u0 += (solution (local_dof_indices[i]) *
+                     fe_values.shape_value_component(i, q_point, 0));
+              v0 += (solution (local_dof_indices[i]) *
+                     fe_values.shape_value_component(i, q_point, 1));
+            }
+          u0 -= 1.0;
```
Comment (tjhei) [inline tests/fe/abf_02.cc]: what does this do, is this the exact gradient? Can you add a comment or clean it up?

File: tests/fe/abf_02.cc [hunk 4]
```diff
@@ -0,0 +1,593 @@
...
+
+
+
+
+
+/*
+ * Check the value of the derivative field.
+ */
+
+void EvaluateDerivative (DoFHandler<3> &dof_handler,
```
Comment (tjhei) [inline tests/fe/abf_02.cc]: what is different between ``integrate_difference`` and this function?

File: tests/fe/abf_02.cc [hunk 5]
```diff
@@ -0,0 +1,593 @@
...
+      else
+        {
+          // Compute eventual sign changes depending on the neighborhood
+          // between two faces.
+          std::vector<double> sign_change (dofs_per_cell, 1.0);
+          const unsigned int dofs_per_face = fe.dofs_per_face;
+          std::vector<unsigned int> face_dof_indices (dofs_per_face);
+
+          /* This code should now be in fe_poly_tensor.cc
+          for (unsigned int f = 2; f < 4; ++f)
```
Comment (tjhei) [inline tests/fe/abf_02.cc]: remove

File: tests/fe/abf_02.cc [hunk 6]
```diff
@@ -0,0 +1,593 @@
...
+              // Precompute the component values
+              for (unsigned int i=0; i < dofs_per_cell; ++i)
+                for (unsigned int comp_i = 0; comp_i < fe.n_components ();
+                     ++comp_i)
+                  {
+                    val_vector[i](comp_i) = sign_change[i] *
+                                            fe_values.shape_value_component(i,point,comp_i);
+                  }
+              // Now eventually switch the sign of some of the ansatzfunctions.
+              // TODO
```
Comment (tjhei) [inline tests/fe/abf_02.cc]: remove

File: tests/fe/abf_02.cc [hunk 7]
```diff
@@ -0,0 +1,593 @@
...
+                  if (fe.get_nonzero_components(i)[comp_i] == true)
+                    {
+                      cell_vector(i) += rhs_values[point](comp_i) *
+                                        val_vector[i](comp_i) * weights[point];
+                    }
+            }
+        }
+      // transfer everything into the
+      // global object. lock the
+      // matrix meanwhile
```
Comment (tjhei) [inline tests/fe/abf_02.cc]: "lock the matrix"?

File: tests/fe/abf_02.cc [hunk 8]
```diff
@@ -0,0 +1,593 @@
...
+  else
+    // no homogeneous boundary values
+    if (project_to_boundary_first == true)
+      // boundary projection required
+      {
+        /*
+                   // set up a list of boundary functions for
+                   // the different boundary parts. We want the
+                   // @p{function} to hold on all parts of the
+                   // boundary
```
Comment (tjhei) [inline tests/fe/abf_02.cc]: remove

File: tests/fe/abf_02.cc [hunk 9]
```diff
@@ -0,0 +1,593 @@
...
+                           dof.max_couplings_between_dofs());
+  DoFTools::make_sparsity_pattern (dof, sparsity);
+  constraints.condense (sparsity);
+
+  SparseMatrix<double> mass_matrix (sparsity);
+  Vector<double> tmp (mass_matrix.n());
+
+  create_mass_matrix (mapping, dof, quadrature, mass_matrix, function, tmp);
+  //create_right_hand_side (mapping, dof, quadrature, function, tmp);
+  //printf ("RHS created\n");
```
Comment (tjhei) [inline tests/fe/abf_02.cc]: remove

File: tests/fe/abf_02.cc [hunk 10]
```diff
@@ -0,0 +1,593 @@
...
+
+  // distribute solution
+  constraints.distribute (vec);
+}
+
+
+
+int main (int /*argc*/, char **/*argv*/)
+{
+  std::ofstream logfile ("output");
```
Comment (tjhei) [inline tests/fe/abf_02.cc]: replace by ``initlog();``

File: tests/fe/abf_02.cc [hunk 11]
```diff
@@ -0,0 +1,593 @@
...
+    {
+      deallog << "Cell " << cell << std::endl;
+      for (unsigned int v=0; v<4; ++v)
+        deallog << "    " << cell->vertex(v) << std::endl;
+    }
+
+
+  FE_ABF<3> fe (0);
+  deallog << "Dofs/cell " << fe.dofs_per_cell
+          << "Dofs/face " << fe.dofs_per_face << std::endl;
```
Comment (tjhei) [inline tests/fe/abf_02.cc]: missing space before ``Dofs``

File: tests/fe/abf_02.cc [hunk 12]
```diff
@@ -0,0 +1,593 @@
...
+int main (int /*argc*/, char **/*argv*/)
+{
+  std::ofstream logfile ("output");
+  deallog << std::setprecision(2);
+  deallog.attach(logfile);
+  deallog.threshold_double(1.e-10);
+
+
+  Triangulation<3> tria_test;
+  GridGenerator::hyper_cube(tria_test);
```
Comment (tjhei) [inline tests/fe/abf_02.cc]: is this test using a single cell?!

File: tests/fe/abf_02.cc [hunk 13]
```diff
@@ -0,0 +1,593 @@
...
+
+  EvaluateDerivative (dof_handler, solution);
+  solution.print (deallog);
+
+  DataOut<3> data_out;
+  data_out.attach_dof_handler (dof_handler);
+  data_out.add_data_vector (solution, "solution");
+  data_out.build_patches (4);
+
+  data_out.write_gnuplot (deallog.get_file_stream());
```
Comment (tjhei) [inline tests/fe/abf_02.cc]: This is quite a bit of output. Did you check visually that it is correct? Needed in the test?

**Comment by bangerth** [comment]
Will take care of this today or tomorrow. The test is actually copied almost verbatim from `abf_01`, so all of the crud is in there as well.

File: tests/fe/abf_02.cc (hunk 13 above)
Comment (bangerth) [inline tests/fe/abf_02.cc]: I didn't check anything -- I only made sure that it actually executes, which it didn't do before (there were at least 2 places where we got exceptions, see the PRs linked to from #2785).

That said, the test does output the L2 error of the L2 projection of a function we ought to be able to represent, and that output is zero.

**Comment by bangerth** [comment]
I've updated the test in most of the requested regards. It is true that much of the test actually duplicates functionality of `VectorTools::project()`. This is because it was copied mostly verbatim from the 2d version in `fe/abf_01`, which has all of the same issues you pointed out here.

I don't know how to handle this. It seems silly to spend a lot of work making tests be as minimal as possible -- that's not where our time is well spent. One can also argue that having multiple ways doing the same thing in the testsuite is not a bad thing (e.g., in this case implementing the projection by hand, rather than calling `VectorTools::project`) because it presumably covers a larger part of the library that may otherwise not be tested as well.

Upshot is, I'm not terribly motivated to work all that much more on this test -- because the `fe/abf_01` test looks even worse. I do think that the test as is provides a benefit. Guidance would be appreciated!

File: tests/fe/abf_02.cc [hunk 14]
```diff
@@ -0,0 +1,552 @@
...
+            {
+              u0 += (solution (local_dof_indices[i]) *
+                     fe_values.shape_value_component(i, q_point, 0));
+              v0 += (solution (local_dof_indices[i]) *
+                     fe_values.shape_value_component(i, q_point, 1));
+            }
+
+          // subtract from it the exact solution
+          u0 -= 1.0;
+          v0 -= 1.0;
```
Comment (tjhei) [inline tests/fe/abf_02.cc]: this looks like it is copy&pasted from the 2d version and won't work correctly in 3d, right?

**Comment by tjhei** [comment]
> I don't know how to handle this.

Yeah, I admit that it is a difficult trade-off to make. I don't insist on having perfectly clean tests, but spending a few minutes now makes our life easier down the road (we have to run/maintain/fix them).

I am still confused why this is only using a single cell. Otherwise I am fine with this.

**Comment by bangerth** [comment]
I think that the test originally was just meant to make sure that the constant function is inside the space. That's not obvious to get right if you have a non-trivial element like the ABF.

I'll look into this some more tonight.

**Comment by bangerth** [comment]
I've updated the one place you pointed out where we compute the 2d error in a 3d setting.

I played with using `VectorTools::project()` and `VectorTools::integrate_difference()` but I get different results that I don't understand. In other words, the test does something that may or may not in itself be correct, but that is nevertheless running through code paths that are worth testing. I don't have the capacity today or in the near future to figure this out -- I do think the ABF element has a good number of issues, but that needs to wait till another day. The intent of the test was to ensure that we can even *create* such an element in 3d, which before wasn't the case.

I'll leave it up to you whether the test is acceptable. It won't hurt my feelings if you say it's not, but that's about the amount of time I can offer at the moment.
